src/tools/run-make-support/src/lib.rs

```diff
@@ -0,0 +1,91 @@
+    /// Initialize the `rmake-test-v2` support context with a snapshot of the current process'
+    /// environment.
+    pub fn init() -> Self {
+        let mut vars = std::env::vars().collect::<HashMap<_, _>>();
+        let ld_lib_path_envvar = vars.get("LD_LIB_PATH_ENVVAR").unwrap();
+        let mut target_rpath_env_path = String::new();
+        target_rpath_env_path.push_str(vars.get("TMPDIR").unwrap());
+        target_rpath_env_path.push(':');
+        target_rpath_env_path.push_str(vars.get(ld_lib_path_envvar).unwrap());
+        vars.insert("TARGET_RPATH_ENV".to_string(), target_rpath_env_path);
+        Self { vars }
+    }
+
+    pub fn tmpdir(&self) -> &str {
+        self.vars.get("TMPDIR").unwrap()
+    }
+
+    pub fn rustc(&self) -> &str {
+        self.vars.get("RUSTC").unwrap()
+    }
+
+    fn run_common(&self, bin_name: String) -> Result<Output, String> {
+        let mut bin_path = PathBuf::new();
+        bin_path.push(self.tmpdir());
+        bin_path.push(&bin_name);
+        let output = Command::new(bin_path)
+            .env("TARGET_RPATH_ENV", self.vars.get("TARGET_RPATH_ENV").unwrap())
+            .env(self.vars.get("LD_LIB_PATH_ENVVAR").unwrap(), {
+                let ld_lib_path_envvar = self.vars.get("LD_LIB_PATH_ENVVAR").unwrap();
+                let mut target_rpath_env_path = String::new();
+                target_rpath_env_path.push_str(self.tmpdir());
+                target_rpath_env_path.push(':');
+                target_rpath_env_path.push_str(self.vars.get("TARGET_RPATH_DIR").unwrap());
+                target_rpath_env_path.push(':');
+                target_rpath_env_path.push_str(self.vars.get(ld_lib_path_envvar).unwrap());
+                target_rpath_env_path
+            })
+            .output()
+            .map_err(|e| format!("{:?}", e))?;
+        Ok(output)
+    }
+
+    /// Run a built binary and make sure it succeeds.
+    pub fn run(&self, bin_name: String) -> Result<Output, String> {
+        let output = self.run_common(bin_name)?;
+        assert!(output.status.success());
+        Ok(output)
+    }
+
+    /// Run a built binary and make sure it fails.
+    pub fn run_fail(&self, bin_name: String) -> Result<Output, String> {
+        let output = self.run_common(bin_name)?;
+        assert!(!output.status.success());
+        Ok(output)
+    }
+}
+
+#[macro_export]
+macro_rules! rustc {
+    ($scx: expr, $args: expr) => {{
+        let sh = $crate::xshell::Shell::new().unwrap();
+        let rustc = $scx.rustc();
+        let args = $args;
+        let mut cmd = $crate::xshell::cmd!(sh, "{rustc}")
+            .arg("--out-dir")
+            .arg($scx.tmpdir())
+            .arg("-L")
+            .arg($scx.tmpdir());
+        for arg in args.split_whitespace() {
+            cmd = cmd.arg(arg);
+        }
+
+        let output = cmd.quiet().output().map_err(|e| format!("{}", e))?;
+        assert!(output.status.success());
+        Ok::<::std::process::Output, String>(output)
```

Comment:
If you use `.unwrap()` or `std::process::exit(1)` instead of `?` above you can directly return the output rather than a `Result`, making tests a bit nicer.